numba/passthru.py

```diff
@@ -0,0 +1,241 @@
+
+
+class PassThruContainerType(PassThruType):
+    def __init__(self):
+        super(PassThruContainerType, self).__init__()
+
+
+pass_thru_container_type = PassThruContainerType()
+
+
+@typeof_impl.register(PassThruContainer)
+def type_pass_thru_container(val, context):
+    return pass_thru_container_type
+
+
+@register_model(PassThruContainerType)
+class PassThruContainerModel(models.StructModel):
+    def __init__(self, dmm, fe_typ):
+        members = [
+            ('container', pass_thru_type),
+            ('wrapped_obj', opaque_pyobject)
+        ]
+        super(PassThruContainerModel, self).__init__(dmm, fe_typ, members)
+
+
+make_attribute_wrapper(PassThruContainerType, 'wrapped_obj', 'wrapped_obj')
+
+
+@unbox(PassThruContainerType)
+def unbox_pass_thru_container_type(typ, obj, context):
+    container = cgutils.create_struct_proxy(typ)(
+        context.context, context.builder
+    )
+
+    container.container = context.unbox(pass_thru_type, obj).value
+
+    wrapped_obj = context.pyapi.object_getattr_string(obj, "obj")
+    context.pyapi.decref(wrapped_obj)
+    container.wrapped_obj = wrapped_obj
+
+    return NativeValue(container._getvalue())
+
+
+@box(PassThruContainerType)
+def box_pass_thru_container_type(typ, val, context):
+    val = cgutils.create_struct_proxy(typ)(
+        context.context, context.builder, value=val
+    )
+
+    return context.box(pass_thru_type, val.container)
+
+
+@lower_builtin(int, types.Opaque)
+def opaque_to_int(context, builder, sig, args):
+    return builder.ptrtoint(args[0], cgutils.intp_t)
+
+
+@type_callable(int)
+def type_opaque_to_int(context):
+    def opaque_to_int_typer(typ):
+        if isinstance(typ, types.Opaque):
+            return types.intp
+
+    return opaque_to_int_typer
+
+
+@lower_builtin(is_, types.Opaque, types.Opaque)
+def opaque_is(context, builder, sig, args):
+    """
+    Implementation for `x is y` for Opaque types. `x is y` iff the pointers
+    are equal
+    """
+    # TODO: would like to use high-level extension here (int(x) == int(y)),
+    #  however currently @overload(is_) does not seem to work, generic impl
+    #  takes precedence
+
+    lhs_type, rhs_type = sig.args
+    # the lhs and rhs have the same type
+    if lhs_type == rhs_type:
+        lhs_ptr = builder.ptrtoint(args[0], cgutils.intp_t)
+        rhs_ptr = builder.ptrtoint(args[1], cgutils.intp_t)
+
+        return builder.icmp_unsigned('==', lhs_ptr, rhs_ptr)
+    else:
+        return cgutils.false_bit
+
+
+@overload(eq)
+def pass_thru_container_eq(x, y):
+    if x is PassThruContainerType():
+        if y is PassThruContainerType():
+            def pass_thru_container_pass_thru_container_eq_impl(x, y):
+                return x.wrapped_obj is y.wrapped_obj
+
+            return pass_thru_container_pass_thru_container_eq_impl
+        else:
+            def pass_thru_container_any_eq_impl(x, y):
+                return False
+
+            return pass_thru_container_any_eq_impl
```

Comment:
What about the case when  `y` is a `PassThruContainerType` but `x` is not?

Isn't the right thing to do here to return `None` and let someone else try in all but the case when both objects are `PassThruContainerType`?

Reply:
Your right, aside from being buggy this is definitely overreaching. Should only handle the case where `x` and `y` are `PassThruContainerType`